acados/ocp_nlp/ocp_nlp_common.c

```diff
@@ -2115,11 +2242,208 @@ void ocp_nlp_embed_initial_value(ocp_nlp_config *config, ocp_nlp_dims *dims,
+        merit_grad += blasfeo_ddot(nv[i], tmp_vec, 0, mem->qp_out->ux + i, 0);
+    }
+    double merit_grad_cost = merit_grad;
+
+    /* dynamics */
+    double merit_grad_dyn = 0.0;
+    for (i=0; i<N; i++)
+    {
+        // get shooting node gap x_next(x_n, u_n) - x_{n+1};
+        tmp_vec = config->dynamics[i]->memory_get_fun_ptr(mem->dynamics[i]);
+
+        /* compute directional derivative of xnext with direction y -> tmp_vec_nxu */
+        blasfeo_dvecse(nx[i], 0.0, &dxnext_dy, 0);
+        for (j = 0; j < nu[i]; j++)
+        {
+            double alpha = BLASFEO_DVECEL(mem->qp_out->ux+i, j);
+            // Su
+            blasfeo_drowex(nx[i], 1.0, mem->qp_in->BAbt, j, 0, &tmp_vec_nxu, 0);
+            blasfeo_daxpy(nx[i], alpha, &tmp_vec_nxu, 0, &dxnext_dy, 0, &dxnext_dy, 0);
+        }
+        for (j = nu[i]; j < nu[i] + nx[i]; j++)
+        {
+            double alpha = BLASFEO_DVECEL(mem->qp_out->ux+i, j);
+            // Sx
+            blasfeo_drowex(nx[i], 1.0, mem->qp_in->BAbt, j, 0, &tmp_vec_nxu, 0);
+            blasfeo_daxpy(nx[i], alpha, &tmp_vec_nxu, 0, &dxnext_dy, 0, &dxnext_dy, 0);
+        }
+        /* add merit gradient contributions depending on sign of shooting gap */
+        for (j = 0; j < nx[i]; j++)
+        {
+            weight = BLASFEO_DVECEL(work->weight_merit_fun->pi+i, j);
+            double deqj_dy = BLASFEO_DVECEL(&dxnext_dy, j) - BLASFEO_DVECEL(mem->qp_out->ux+(i+1), nu[i+1]+j);
+            {
+                if (BLASFEO_DVECEL(tmp_vec, j) > 0)
+                {
+                    merit_grad_dyn += weight * deqj_dy;
+                    // printf("\ndyn_contribution +%e, weight %e, deqj_dy %e, i %d, j %d", weight * deqj_dy, weight, deqj_dy, i, j);
+                }
+                else
+                {
+                    merit_grad_dyn -= weight * deqj_dy;
+                    // printf("\ndyn_contribution %e, weight %e, deqj_dy %e, i %d, j %d", -weight * deqj_dy, weight, deqj_dy, i, j);
+                }
+            }
+        }
+    }
+
+    /* inequality contributions */
+    // NOTE: slack bound inequalities are not considered here.
+    // They should never be infeasible. Only if explicitly initialized infeasible from outside.
+    int constr_index, slack_index_in_ux, slack_index;
+    ocp_qp_dims* qp_dims = mem->qp_in->dim;
+    int *nb = qp_dims->nb;
+    int *ng = qp_dims->ng;
+    int *ns = qp_dims->ns;
+    double merit_grad_ineq = 0.0;
+    double slack_step;
+
+    for (i=0; i<=N; i++)
+    {
+        tmp_vec = config->constraints[i]->memory_get_fun_ptr(mem->constraints[i]);
+        int *idxb = mem->qp_in->idxb[i];
+        if (ni[i] > 0)
+        {
+            // printf("constraints fun = \n");
+            // blasfeo_print_exp_dvec(2*ni[i], tmp_vec, 0);
+            // print_ocp_qp_in(mem->qp_in);
+            // for (j = 0; j < 2 * ni[i]; j++) // 2 * ni
+            for (j = 0; j < 2 * (nb[i] + ng[i]); j++) // 2 * ni
```

Comment:
Since everything is two sided, I would rather loop with j from 0 to nb[i]+ng[i] and process the upper and the lower bounds at once merged in the same loop.
This would simplify indexing and give extra independent work for each loop iteration.

Reply:
I think here, there is no performance difference, since this is just done for either upper or lower constraints or none of them.
So, I just left a note https://github.com/acados/acados/pull/772/commits/c1f10500125b8bdd6f86a44c745a22957e58cae2